whylog/constraints/__init__.py

```
@@ -2,42 +2,177 @@
-    def __init__(self, intervals):
-        pass
+    @classmethod
+    def get_groups_count(cls):
+        return super(TimeConstraint, cls).get_groups_count()
 
+    @classmethod
+    def verify(cls, param_dict, group_contents):
+        #TODO
+        raise NotImplementedError
 
-class AnyValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
 
+class IdenticalConstraint(AbstractConstraint):
+    """
+    Contents of groups must be identical.
+    """
 
-class DifferentValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    TYPE = ConstraintType.IDENTICAL
 
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = None
 
-class ConstIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    def __init__(self, param_dict, groups):
+        """
+        I.e:
+        IdenticalConstraint(
+            dict(),
+            [(1, 2), (2, 4)]
+        )
+        """
+        super(IdenticalConstraint, self).__init__(param_dict, groups)
+
+    def convert_to_user_constraint_intent(self):
+        return super(IdenticalConstraint, self).convert_to_user_constraint_intent()
 
+    @classmethod
+    def get_param_names(cls):
+        return []
 
-class ValueDeltaIntervals(AbstractConstraint):
-    def __init__(self, interval_lower, interval_greater, min_delta=None, max_delta=None):
+    @classmethod
+    def get_groups_count(cls):
+        return super(IdenticalConstraint, cls).get_groups_count()
+
+    @classmethod
+    def verify(cls, param_dict, group_contents):
         """
-        Sets Minimum and maximum difference between values of params (if values are numbers).
+        I.e:
+        - verify({}, ['comp1', 'comp1', 'comp1']) should pass
+        - verify({}, ['comp1', 'hello', 'comp1']) should raise error
         """
-        pass
+        if not len(set(group_contents)) == 1:
+            raise ConstraintVerificationError(cls.TYPE, param_dict, group_contents)
```

Why you raise error when verification failed, instead of just returning False?